### PR TITLE
Crypto: add native implementation

### PR DESCRIPTION
**Pull request by semuxdev**
Resolves #760 #352 

Benchmark (ubuntu 16.04 LTS, Intel® Core™ i7-8700K CPU @ 3.70GHz × 12 ):
```
23:35:49.627 DEBUG    NativeTest       Ed25519 verify native: 1021ms
23:35:52.467 DEBUG    NativeTest       Ed25519 verify java: 2839ms
23:35:53.113 DEBUG    NativeTest       Ed25519 sign native: 449ms
23:35:54.539 DEBUG    NativeTest       Ed25519 sign java: 1426ms
23:35:54.638 DEBUG    NativeTest       Blake2b native: 13ms
23:35:54.705 DEBUG    NativeTest       Blake2b java: 67ms
```

- [x] Add JNI implementation
- [x] Add unit tests
- [x] Add pre-built binaries
- [ ] Integrate into Maven

### CODE REVIEW

File: src/main/java/org/semux/crypto/Key.java [hunk 1]
```diff
@@ -162,19 +167,23 @@ public Signature sign(byte[] msgHash) {
...
      *            signature
      * @return True if the signature is valid, otherwise false
      */
-    public static boolean verify(byte[] msgHash, Signature signature) {
-        if (msgHash != null && signature != null) { // avoid null pointer exception
+    public static boolean verify(byte[] message, Signature signature) {
+        if (message != null && signature != null) { // avoid null pointer exception
             try {
-                EdDSAEngine engine = new EdDSAEngine();
-                engine.initVerify(PublicKeyCache.computeIfAbsent(signature.getPublicKey()));
```
Comment (orogvany) [inline src/main/java/org/semux/crypto/Key.java]: Is there some corollary to caching the public key result for the native lib, or any optimization possible here?

File: src/main/java/org/semux/crypto/Native.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    protected static boolean enabled = false;
+
+    /**
+     * Initializes the native libraries
+     */
+    protected static void init() {
+        SystemUtil.OsName os = SystemUtil.getOsName();
+        switch (os) {
+        case LINUX:
+            // TODO: add 32-bit linux build
```
Comment (orogvany) [inline src/main/java/org/semux/crypto/Native.java]: errant todo?

File: src/test/java/org/semux/crypto/NativeTest.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            key.verify(data, sig);
+        }
+
+        // native
+        Instant start = Instant.now();
+        for (int i = 0; i < repeat; i++) {
+            Native.ed25519_verify(data, sigNative, PUBLIC_KEY);
+        }
+        Instant end = Instant.now();
+        System.out.println("Ed25519 verify native: " + Duration.between(start, end).toMillis() + "ms");
```
Comment (orogvany) [inline src/test/java/org/semux/crypto/NativeTest.java]: logger

File: src/main/native/crypto/org_semux_crypto_Native.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    env->SetByteArrayRegion(result, 0, sizeof(hash), (const jbyte*)hash);
+    return result;
+}
+
+JNIEXPORT jbyteArray JNICALL Java_org_semux_crypto_Native_ed25519_1sign
+(JNIEnv *env, jclass cls, jbyteArray msg, jbyteArray sk)
+{
+    // basic checks
+    if (msg == NULL || sk == NULL || env->GetArrayLength(sk) != crypto_sign_ed25519_SECRETKEYBYTES) {
+        return NULL;
```
Comment (cryptokat) [inline src/main/native/crypto/org_semux_crypto_Native.cpp]: It is suggested to throw an exception here to provide a consistent behaviour with the java implementation.

File: src/main/native/crypto/org_semux_crypto_Native.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    // create ed25519 signature
+    jsize msg_size = env->GetArrayLength(msg);
+    jbyte *msg_ptr = env->GetByteArrayElements(msg, NULL);
+    jbyte *sk_ptr = env->GetByteArrayElements(sk, NULL);
```
Comment (cryptokat) [inline src/main/native/crypto/org_semux_crypto_Native.cpp]: `GetByteArrayRegion` is recommended over `GetByteArrayElements` to reduce overhead.
Ref: https://developer.android.com/training/articles/perf-jni.html#region_calls

File: src/main/native/crypto/org_semux_crypto_Native.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    // verify ed25519 signature
+    jsize msg_size = env->GetArrayLength(msg);
+    jbyte *msg_ptr = env->GetByteArrayElements(msg, NULL);
+    jbyte *sig_ptr = env->GetByteArrayElements(sig, NULL);
+    jbyte *pk_ptr = env->GetByteArrayElements(pk, NULL);
+    jboolean result = crypto_sign_ed25519_verify_detached((const unsigned char *)sig_ptr,
+        (const unsigned char *)msg_ptr, msg_size, (const unsigned char *)pk_ptr) == 0;
+    env->ReleaseByteArrayElements(msg, pk_ptr, 0);
+    env->ReleaseByteArrayElements(msg, sig_ptr, 0);
```
Comment (cryptokat) [inline src/main/native/crypto/org_semux_crypto_Native.cpp]: Typos here. They should be:
```
env->ReleaseByteArrayElements(pk, pk_ptr, 0);
env->ReleaseByteArrayElements(sig, sig_ptr, 0);
```

File: src/main/native/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Semux Native Library
+
+## Build
+
+```
+mkdir build
+cd build
+cmake ..
+make
+```
```
Comment (cryptokat) [inline src/main/native/README.md]: There's no instruction about how cross-compilation can be performed for a new release.

Comment (cryptokat) [reply src/main/native/README.md]: Tools like [nar-maven-plugin](http://maven-nar.github.io/index.html) can be useful

Comment (semuxdev) [reply src/main/native/README.md]: This is cool. I will investigate it!

File: src/main/native/README.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+## Build on macOS
+
+Prerequisites:
+1. Build and install libsodium
+
+Build:
+```
+mkdir build && cd build
```
Comment (orogvany) [inline src/main/native/README.md]: odd.  When attempting this, I get
mac:semux chris.orogvany$ mkdir build && cd build
mac:build chris.orogvany$ cmake ..
CMake Error: The source directory "/Users/USER/Desktop/semux2/semux" does not appear to contain CMakeLists.txt.

**Comment by semuxdev** [comment]
All requested changes have been applied, except maven integration.

Request for a second round review!

**Comment by orogvany** [comment]
Still need to update LICENSE_3RD as libsodium ISC license requires attribution, and we're packaging it with our distro, otherwise, ship it!

**Pull request merged**